crates/btcio/src/reader/query.rs

```diff
@@ -279,7 +283,56 @@ async fn poll_for_new_blocks<R: Reader>(
         };
     }
 
-    Ok(events)
+    Ok(())
+}
+
+/// Watches the status channel for chainstate which has the finalized state matching as referred by
+/// checkpoint. If it is found, returns `true`, otherwise returns `false`. `false` is returned when
+/// there is a different chainstate than indicated by checkpoint for given slot.
+async fn wait_until_checkpoint_accepted_or_rejected<R: Reader>(
+    ctx: &ReaderContext<R>,
+    checkpoint: &SignedCheckpoint,
+) -> anyhow::Result<bool> {
+    let batch_info = checkpoint.checkpoint().batch_info();
+    let l2_commt = checkpoint.checkpoint().batch_info().l2_range.1;
+    let ckpt_epoch = batch_info.epoch();
+
+    let s = info_span!("wait_checkpoint_accepted", %ckpt_epoch, last_blkid=%l2_commt.blkid());
+    let _ = s.enter();
+
+    info!("Waiting for checkpoint to be accepted");
+    let mut rx = ctx.status_channel.subscribe_chain_sync();
+
+    let timeout_duration = Duration::from_secs(5); // TODO: better, clear duration
+
+    loop {
+        if let Some(chstate_update) = (*rx.borrow_and_update()).clone() {
+            let fin_epoch = chstate_update.new_status().finalized_epoch;
+
+            let same_slot = l2_commt.slot() == fin_epoch.last_slot();
+            let same_blkid = l2_commt.blkid() == fin_epoch.last_blkid();
+
+            if same_slot && same_blkid {
+                return Ok(true);
+            } else if same_slot && !same_blkid {
+                info!("Checkpoint not accepted: different block id for same slot finalized");
+                return Ok(false);
+            } else if l2_commt.slot() < fin_epoch.last_slot() {
+                // We've been looking for stale checkpoint to be accepted.
+                info!("Checkpoint not accepted: higher slot finalized than the slot waited for");
+                return Ok(false);
+            }
+        }
+        // Wait for some change to happen
+        tokio::select! {
+            _ = rx.changed() => {}
+            // NOTE: what timeout to wait for here?
+            _ = tokio::time::sleep(timeout_duration) => {
+                error!("Timed out waiting for chainstate update.");
+                return Err(anyhow!("reader:query:wait_for_checkpoint_accept: Timed out waiting for chainstate update"));
+            }
+        }
+    }
```

Comment:
Does it make sense to error out in the loop this way?

Make sure that this usage of `watch` is correct.  Perhaps `.wait_for` might be more applicable?  You could wrap a call to that in the timeout and probably not need the loop around all of this.

Also should be a more structured error type instead of `anyhow!`.

Reply:
Using `.wait_for` is not suitable here because we not only need to wait for something to happen but also exit when something else happens. In this particular case we are waiting for checkpoint X at a height to be accepted but we should also wait until some other checkpoint at the height gets accepted and return boolean accordingly.

> Also should be a more structured error type instead of anyhow!.
Not sure if it needs to be more structured since this is not a library module and we just exit the client when the reader fails for some reason.